code/parse/sexp.cpp

```diff
@@ -29817,6 +31175,9 @@ const char *CTEXT(int n)
 
 	// Goober5000 - MWAHAHAHAHAHAHAHA!  Thank you, Volition programmers!  Without
 	// the CTEXT wrapper, when-argument would probably be infeasibly difficult to code.
+	//
+	// Karajorma - Due to the addition of the container code, coders shouldn't call CTEXT twice from the same function for the same sexp node.
```

Comment:
Why shouldn't `CTEXT()` be called twice on the same node? Is it because of text replacement that, in the case of lists, can modify the list?

Reply:
IMO, ideally `CTEXT()` should be an idempotent operation (i.e., calling it multiple times should have the same effect as calling it once).

If we cache in a `sexp_node`'s `sexp_cached_data` field the result of calling `CTEXT()` on it, shouldn't that ensure that `CTEXT()` remains idempotent?

Reply:
I probably should have made that comment clearer. I actually did mean to have a discussion about it at some point.

CTEXT has never been idempotent. It's just that the containers stuff reveals that. 

For the most part this isn't really an issue. The comment relates to a bad habit a few coders had of writing code for sexps like this 

something  CTEXT(n) 
other operations
something CTEXT(n)

instead of simply buffering the results of the first call to CTEXT. For the most part this was fine but it was banking on variables not changing (which to be fair, doesn't happen with any of the SEXPs I'm aware of). 

The big problem comes if you use the remove functions for a container. The second call to CTEXT(n) will return a different value since the original value in the container has now been removed. The solution is to not be lazy and just buffer the results of the first call to CTEXT since you shouldn't be calling such a computationally expensive function a second time anyway. 

The comment should warn people to not call CTEXT twice for the same node within a function as the contents of that node may have changed. 



As for the sexp_cached_data, I don't remember that existing when I wrote the container code. Now that I've looked at it, I don't think we could use it for that. In fact, it is actually the source of a bug in the container code since the game checks for variables or arguments on the grounds they might change but NOT containers. 

The if (!(Sexp_nodes[node].type & SEXP_FLAG_VARIABLE) && !(Sexp_nodes[node].flags & SNF_SPECIAL_ARG_IN_NODE)) style checks should probably be replaced with a function that checks if the node is a variable, container or argument.